src/hotspot/share/utilities/hashtable.cpp

```diff
@@ -45,56 +44,21 @@
 #include "utilities/hashtable.inline.hpp"
 #include "utilities/numberSeq.hpp"
 
-
 // This hashtable is implemented as an open hash table with a fixed number of buckets.
 
-template <MEMFLAGS F> BasicHashtableEntry<F>* BasicHashtable<F>::new_entry_free_list() {
-  BasicHashtableEntry<F>* entry = NULL;
-  if (_free_list != NULL) {
-    entry = _free_list;
-    _free_list = _free_list->next();
-  }
-  return entry;
-}
+// Hashtable entry allocates in the C heap directly.
 
-// HashtableEntrys are allocated in blocks to reduce the space overhead.
 template <MEMFLAGS F> BasicHashtableEntry<F>* BasicHashtable<F>::new_entry(unsigned int hashValue) {
-  BasicHashtableEntry<F>* entry = new_entry_free_list();
-
-  if (entry == NULL) {
-    if (_first_free_entry + _entry_size >= _end_block) {
-      int block_size = MAX2((int)_table_size / 2, (int)_number_of_entries); // pick a reasonable value
-      block_size = clamp(block_size, 2, 512); // but never go out of this range
-      int len = round_down_power_of_2(_entry_size * block_size);
-      assert(len >= _entry_size, "");
-      _first_free_entry = NEW_C_HEAP_ARRAY2(char, len, F, CURRENT_PC);
-      _entry_blocks.append(_first_free_entry);
-      _end_block = _first_free_entry + len;
-    }
-    entry = (BasicHashtableEntry<F>*)_first_free_entry;
-    _first_free_entry += _entry_size;
-  }
-
+  BasicHashtableEntry<F>* entry = ::new (NEW_C_HEAP_ARRAY(char, this->entry_size(), F))
+                                        BasicHashtableEntry<F>(hashValue);
   assert(_entry_size % HeapWordSize == 0, "");
```

Comment:
Is this assert still needed? What's its purpose?

Reply:
Yeah, seems a bit strange.  Of course the entry is a heapword size.  I'll remove the assert.